### PR TITLE
schedule: cache the unprocessed region due to limit

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve?

Use a LRU cache to save the unprocessed regions due to the limit restriction.

### What is changed and how it works?

This PR adds a `regionWaitingList` in `CheckerController`.
To make it more clear, we mainly have two kinds of situations that need to add the region into the `regionWaitingList`:
1. when the operator is created but the schedule limit reaches the threshold
2. the store limit inside checker reaches the threshold

### Check List

Tests

- Unit test

### CODE REVIEW

**Review by HunDunDM — APPROVED** [review]

**Review by lhy1024 — COMMENTED** [review]
LGTM. will we merge `checkSuspectRegion` and `checkWaitingRegions`？And will we use different cache according to different checkers?

**Comment by BusyJay** [comment]
Any progress?

**Comment by disksing** [comment]
@BusyJay  test result not good enough now, we are still try to figure it out.

File: server/cluster/cluster_stat.go [hunk 1]
```diff
@@ -14,11 +14,12 @@
 package cluster
 
 import (
-	"github.com/tikv/pd/pkg/movingaverage"
 	"strings"
 	"sync"
 	"time"
 
+	"github.com/tikv/pd/pkg/movingaverage"
+
```
Comment (lhy1024) [inline server/cluster/cluster_stat.go:22]: unnecessary empty line?

**Comment by nolouch** [comment]
Does this PR ready to review？

**Comment by rleungx** [comment]
> Does this PR ready to review？

You can review it right now

**Comment by BusyJay** [comment]
PTAL

File: server/schedule/checker_controller.go [hunk 2]
```diff
@@ -34,64 +38,80 @@ type CheckerController struct {
...
 	return &CheckerController{
 		cluster:           cluster,
 		opts:              cluster.GetOpts(),
 		opController:      opController,
 		learnerChecker:    checker.NewLearnerChecker(cluster),
-		replicaChecker:    checker.NewReplicaChecker(cluster),
-		ruleChecker:       checker.NewRuleChecker(cluster, ruleManager),
+		replicaChecker:    checker.NewReplicaChecker(cluster, regionWaitingList),
+		ruleChecker:       checker.NewRuleChecker(cluster, ruleManager, regionWaitingList),
 		mergeChecker:      checker.NewMergeChecker(ctx, cluster),
```
Comment (lhy1024) [inline server/schedule/checker_controller.go:55]: will we add `regionWaitingList` for `mergeChecker`?

File: server/schedule/checker_controller.go [hunk 3]
```diff
@@ -34,64 +38,80 @@ type CheckerController struct {
...
 		jointStateChecker: checker.NewJointStateChecker(cluster),
+		regionWaitingList: regionWaitingList,
 	}
 }
 
 // CheckRegion will check the region and add a new operator if needed.
-func (c *CheckerController) CheckRegion(region *core.RegionInfo) (bool, []*operator.Operator) { //return checkerIsBusy,ops
+func (c *CheckerController) CheckRegion(region *core.RegionInfo) []*operator.Operator { //return checkerIsBusy,ops
```
Comment (lhy1024) [inline server/schedule/checker_controller.go:62]: why to remove `CheckerIsBusy`?

File: server/schedule/checker_controller.go (hunk 2 above)
Comment (rleungx) [reply server/schedule/checker_controller.go:55]: No, for now, we don't save the regions which need to be merged into the cache.

File: server/schedule/checker_controller.go (hunk 3 above)
Comment (rleungx) [reply server/schedule/checker_controller.go:62]: Since it is not used anymore

File: server/schedule/checker_controller.go [hunk 4]
```diff
@@ -34,64 +38,80 @@
...
 		}
-		if opController.OperatorCount(operator.OpReplica) < c.opts.GetReplicaScheduleLimit() {
-			checkerIsBusy = false
-			if op := c.replicaChecker.Check(region); op != nil {
-				return checkerIsBusy, []*operator.Operator{op}
+		if op := c.replicaChecker.Check(region); op != nil {
+			if opController.OperatorCount(operator.OpReplica) < c.opts.GetReplicaScheduleLimit() {
+				return []*operator.Operator{op}
 			}
+			c.regionWaitingList.Put(region.GetID(), nil)
```
Comment (HunDunDM) [inline server/schedule/checker_controller.go:86]: It seems that it can be put inside the checker or outside the checker. Can you explain the design?

**Review by HunDunDM — COMMENTED** [review]
rest LGTM

File: server/schedule/operator_controller.go [hunk 5]
```diff
@@ -123,10 +123,12 @@ func (oc *OperatorController) Dispatch(region *core.RegionInfo, source string) {
 		case operator.SUCCESS:
 			oc.pushHistory(op)
 			if oc.RemoveOperator(op) {
+				operatorWaitCounter.WithLabelValues(op.Desc(), "promote-success").Inc()
```
[the remark this answers is not in the document]
Comment (rleungx) [reply server/schedule/operator_controller.go:126]: They don't have any relationship.

File: server/schedule/operator_controller.go [hunk 6]
```diff
@@ -298,6 +303,7 @@ func (oc *OperatorController) AddWaitingOperator(ops ...*operator.Operator) int
 	}
 
 	oc.Unlock()
+	operatorWaitCounter.WithLabelValues(ops[0].Desc(), "promote-add").Inc()
```
[the remark this answers is not in the document]
Comment (rleungx) [reply server/schedule/operator_controller.go:306]: ditto

File: server/schedule/checker_controller.go (hunk 2 above)
Comment (lhy1024) [reply server/schedule/checker_controller.go:55]: So will it affect the speed of merge?

File: server/schedule/checker_controller.go (hunk 3 above)
Comment (lhy1024) [reply server/schedule/checker_controller.go:62]: then we should remove the comment.

File: server/schedule/checker_controller.go (hunk 2 above)
Comment (lhy1024) [reply server/schedule/checker_controller.go:55]: Maybe will it be better to provide a cache for each checker

**Review by Yisaer — APPROVED** [review]
LGTM

**Pull request merged**